contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -944,6 +944,53 @@ CAPIStatus Filter::log(uint32_t level, absl::string_view message) {
   PANIC_DUE_TO_CORRUPT_ENUM;
 }
 
+CAPIStatus Filter::setDynamicMetadata(std::string filter_name, std::string key,
+                                      absl::string_view bufStr) {
+  Thread::LockGuard lock(mutex_);
+  if (has_destroyed_) {
+    ENVOY_LOG(debug, "golang filter has been destroyed");
+    return CAPIStatus::CAPIFilterIsDestroy;
+  }
+
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+
+  if (!state.isThreadSafe()) {
+    auto weak_ptr = weak_from_this();
+    state.getDispatcher().post([this, &state, weak_ptr, filter_name, key, bufStr] {
+      ASSERT(state.isThreadSafe());
+      // TODO: do not need lock here, since it's the work thread now.
+      Thread::ReleasableLockGuard lock(mutex_);
+      if (!weak_ptr.expired() && !has_destroyed_) {
+        lock.release();
+        // it's safe reuse bufStr since Go will wait until C callback.
```

Comment:
Oh, sorry, I think this is wrong.
Go only waits for the CAPIStatus return value (line 975), while this will be delivered to run in worker thead async, that means Go may already free the slice buffer.
Maybe we need to copy bufStr here, similar to:
https://github.com/envoyproxy/envoy/blob/7ad71903cc61ede16aad20894f8aeb9f2cd2ed1d/contrib/golang/filters/http/source/golang_filter.cc#L722-L723

Reply:
ohhhh yeah that makes sense now. Thanks!